pkg/kubernetes/kubernetes.go

```diff
@@ -132,3 +156,62 @@ func (m *Mixin) handleOutputs(outputs []KubernetesOutput) error {
 	}
 	return nil
 }
+
+func installKubectlClient(m *Mixin, version string) error {
+
+	url := fmt.Sprintf("https://storage.googleapis.com/kubernetes-release/release/%s/bin/linux/amd64/kubectl", version)
+
+	// Fetch archive from url
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to construct GET request for fetching kubectl client binary")
+	}
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return errors.Wrapf(err, "failed to download kubectl client binary via url: %s", url)
+	}
+	defer res.Body.Close()
+
+	// Create a temp dir
+	tmpDir, err := m.FileSystem.TempDir("", "tmp")
+	if err != nil {
+		return errors.Wrap(err, "unable to create a temporary directory for downloading the kubectl client binary")
+	}
+	defer os.RemoveAll(tmpDir)
+
+	// Create the local archive
+	kubectlBinPath, err := m.FileSystem.Create(filepath.Join(tmpDir, "kubectl"))
+	if err != nil {
+		return errors.Wrap(err, "unable to create a local file for the kubectl client binary")
+	}
+
+	// Copy response body to local archive
+	_, err = io.Copy(kubectlBinPath, res.Body)
+	if err != nil {
+		return errors.Wrap(err, "unable to copy the kubectl client binary to the local archive file")
+	}
+
+	// Move the kubectl binary into the appropriate location
+	binPath := "/usr/local/bin/kubectl"
+	err = m.FileSystem.Rename(fmt.Sprintf("%s", kubectlBinPath), binPath)
+	if err != nil {
+		return errors.Wrapf(err, "unable to install the kubectl client binary to %q", binPath)
+	}
+	return nil
+}
+
+func getKubectlApiServerVersion(m *Mixin) (string, error) {
+	var stderr bytes.Buffer
+
+	cmd := m.NewCommand("kubectl", "api-versions")
```

Comment:
This is the wrong command to get the semver version of the kubernetes api server. What you have here will output 

```
$ kubectl api-versions
admissionregistration.k8s.io/v1beta1
apiextensions.k8s.io/v1beta1
apiregistration.k8s.io/v1
apiregistration.k8s.io/v1beta1
apps/v1
```

Instead try this and grab the `serverVersion.gitversion`:

```
$ kubectl version -o json
{
  "clientVersion": {
    "major": "1",
    "minor": "18",
    "gitVersion": "v1.18.3",
    "gitCommit": "2e7996e3e2712684bc73f0dec0200d64eec7fe40",
    "gitTreeState": "clean",
    "buildDate": "2020-05-21T14:51:23Z",
    "goVersion": "go1.14.3",
    "compiler": "gc",
    "platform": "darwin/amd64"
  },
  "serverVersion": {
    "major": "1",
    "minor": "15",
    "gitVersion": "v1.15.10",
    "gitCommit": "ba1ce4122d67b0f7399a41afbbd8a7258d22ecb8",
    "gitTreeState": "clean",
    "buildDate": "2020-04-21T00:13:25Z",
    "goVersion": "go1.12.12",
    "compiler": "gc",
    "platform": "linux/amd64"
  }
}
```

Reply:
You are right that was totally the wrong command😅 . I changed it the to right command and added a type to parse the json output. But I realize that in order to execute `kubectl version -o json` we need an already present client . I wonder how we can pass the kubectl client to execute this command at runtime

Reply:
The mixin always installs a kubectl client during `porter build`. So you can rely on it being in the invocation image when you call this command to detect the cluster version.